### PR TITLE
feat(cloud): Add cloud notifications cli for list, set, update and delete notifications

### PR DESCRIPTION
**Pull request by leoroese**
Add cli for notifications and documentation

Per discussion here
https://github.com/meltano/internal-product/issues/220#issuecomment-1561987521

Cloud backend PR
https://github.com/meltano/infra/pull/1730

Event payload PR
https://github.com/meltano/infra/pull/2074

Closes: https://github.com/meltano/meltano/issues/8030

This allows users to 
1. list all configured notifications for organization or project
2. set new notifications with filters
3. update a notifications recipient or status
4. Delete notifications 

### Testing

example use commands
**list notifications**
`meltano cloud config notification list`

**set notification**
`
meltano cloud config notification set webhook --recipient https://test+melty --filter '{"events": ["job_run"], "status": ["failed", "succeeded"]}' \
--filter '{"status": ["failed"]}'
`

**update notifcation url or email**
`meltano cloud config notification update webhook --recipient http://test --new http://test+melty --status inactive`

**delete notification**
`meltano cloud config notification delete webhook --recipient http://test`

### CODE REVIEW

File: src/cloud-cli/meltano/cloud/cli/config.py
Comment (leoroese) [inline src/cloud-cli/meltano/cloud/cli/config.py:430]: Having this set as a notification group command ends up breaking pythons built in set(). I tried to follow similar semantics from 
https://docs.meltano.com/reference/command-line-interface/#config

2 options is
1. Do similar and use set comprehension
2. Import set as a new name like pyset
3. Rename this command to put or something else

Curious about thoughts on this

Comment (WillDaSilva) [reply src/cloud-cli/meltano/cloud/cli/config.py:430]: @leoroese You can rename the function to whatever you want, and then specify the name it should have in the CLI as a param to the `group` function

File: docs/docs/cloud/cloud-cli.mdx
Comment (tayloramurphy) [inline docs/docs/cloud/cloud-cli.mdx:86]: @leoroese presumably all of these are case-sensitive, correct? Can we add a line indicating as much?

File: docs/docs/cloud/cloud-cli.mdx
Comment (tayloramurphy) [inline docs/docs/cloud/cloud-cli.mdx:101]: Perhaps in a future iteration we can add an `alias` or `name` parameter to the notificaiton. It feels a little weird to reference an existing notification by where it's going, but not weird enough to block. It gets the feature done 👍

File: docs/docs/cloud/cloud-cli.mdx
Comment (tayloramurphy) [inline docs/docs/cloud/cloud-cli.mdx:62]: Can we put an example output here of what users should expect?

File: docs/docs/cloud/concepts.md
Comment (tayloramurphy) [inline docs/docs/cloud/concepts.md:91]:
```suggestion
A Meltano Cloud notification is a response to Meltano Cloud event.
```

**Review by tayloramurphy — CHANGES_REQUESTED** [review]

File: docs/docs/cloud/cloud-cli.mdx
Comment (leoroese) [reply docs/docs/cloud/cloud-cli.mdx:86]: Oh yeah good call

**Review by tayloramurphy — APPROVED** [review]

**Comment by leoroese** [comment]
Could I get a review from @meltano/engineering on this if someone has the time?

File: docs/docs/cloud/cloud-cli.mdx
Comment (edgarrmondragon) [inline docs/docs/cloud/cloud-cli.mdx:69]:
```suggestion
$ meltano cloud config notification list
╭─────────────────────┬─────────┬──────────╮
│ Recipient           │ Type    │ Status   │
├─────────────────────┼─────────┼──────────┤
│ https://melty       │ webhook │ active   │
│ dragon@meltano.com  │ email   │ active   │
╰─────────────────────┴─────────┴──────────╯
```

File: docs/docs/cloud/cloud-cli.mdx
Comment (WillDaSilva) [inline docs/docs/cloud/cloud-cli.mdx:51]: Thoughts on renaming the command to `notifications`, i.e. make it plural?

File: src/meltano/cloud/cli/config.py
Comment (edgarrmondragon) [inline src/meltano/cloud/cli/config.py:92]:
```suggestion
```

File: src/meltano/cloud/cli/config.py
Comment (WillDaSilva) [inline src/meltano/cloud/cli/config.py:645]:
```suggestion
                f"Please provide old {type} recipient for notification to be updated"
```

File: src/meltano/cloud/cli/config.py
Comment (edgarrmondragon) [inline src/meltano/cloud/cli/config.py:87]:
```suggestion
        """Build a notification url.
```

File: src/meltano/cloud/cli/config.py
Comment (edgarrmondragon) [inline src/meltano/cloud/cli/config.py:154]:
```suggestion
        """Update the notification key.
```

File: src/meltano/cloud/cli/config.py
Comment (WillDaSilva) [inline src/meltano/cloud/cli/config.py:718]: It's not clear what should be specified when given this prompt. The message should make it clear that we're looking for one of the items in the "recipient" column of the `notification list` command. I'm not sure how to concisely phrase that.

**Review by WillDaSilva — APPROVED** [review]

File: docs/docs/cloud/cloud-cli.mdx
Comment (WillDaSilva) [reply docs/docs/cloud/cloud-cli.mdx:51]: There was a bit discussion about this a long time ago where we decided that we wanted to go with plural by default, but then had a bunch of exceptional reasons for making particular commands singular. Now it's a mix :sweat_smile:

I'm fine either way, and don't want to hold up this PR.

File: src/meltano/cloud/cli/config.py
Comment (edgarrmondragon) [inline src/meltano/cloud/cli/config.py:741]: Ditto. Should these be assigned to variables?

**Review by edgarrmondragon — CHANGES_REQUESTED** [review]
A few changes

File: docs/docs/cloud/cloud-cli.mdx
Comment (leoroese) [reply docs/docs/cloud/cloud-cli.mdx:69]: Oh yeah that looks much better!

File: src/meltano/cloud/cli/config.py
Comment (leoroese) [reply src/meltano/cloud/cli/config.py:585]: The reason I decided to not assign them to variables is because these functions aren't being referenced anywhere else. 

It currently works due to how Click works. It's my understanding that Click registers these commands behind the scenes so they're immediately usable as commands without the need for a local handle.

Same answer for all 3 other comments related to this. I am indifferent about whether to assign these to a variable, either way would work.

File: docs/docs/cloud/cloud-cli.mdx
Comment (tayloramurphy) [inline docs/docs/cloud/cloud-cli.mdx:53]:
```suggestion
The `notification` subcommand provides a way for managing your [Meltano Cloud notifications](concepts#meltano-cloud-notifications).
```

Technically a subcommand, right?

File: docs/docs/cloud/cloud-cli.mdx
Comment (tayloramurphy) [inline docs/docs/cloud/cloud-cli.mdx:60]: @leoroese I know we're deep into this PR but how hard is it to show the filter on the notification list? 

My process was I did a new interactive notification but all it asked for was my email. Then when I do `list` I just see the active notification but with no idea what notifications I'm going to get.

File: docs/docs/cloud/cloud-cli.mdx
Comment (tayloramurphy) [inline docs/docs/cloud/cloud-cli.mdx:92]: What is the default filter? If I don't set anything what does the payload look like?

File: docs/docs/cloud/cloud-cli.mdx
Comment (tayloramurphy) [inline docs/docs/cloud/cloud-cli.mdx:127]: @leoroese what happens if I have multiple notifications going to the same recipient? When I run the delete does it delete them all?

File: docs/docs/cloud/cloud-cli.mdx
Comment (tayloramurphy) [inline docs/docs/cloud/cloud-cli.mdx:98]: Can we have a few examples of what we think will be the most common filters? 

* job success
* job failure

any others?

**Review by tayloramurphy — CHANGES_REQUESTED** [review]

File: docs/docs/cloud/cloud-cli.mdx
Comment (leoroese) [reply docs/docs/cloud/cloud-cli.mdx:53]: Yes!

File: docs/docs/cloud/cloud-cli.mdx
Comment (leoroese) [reply docs/docs/cloud/cloud-cli.mdx:60]: It's currently being returned, it would be super easy to display

File: docs/docs/cloud/cloud-cli.mdx
Comment (leoroese) [reply docs/docs/cloud/cloud-cli.mdx:92]: Default filter is just an empty list meaning that there are no filters and you will always receive the notification. I believe payload just includes `filter: []`, when returned from the list, it just shows an empty list as well

File: docs/docs/cloud/cloud-cli.mdx
Comment (leoroese) [reply docs/docs/cloud/cloud-cli.mdx:127]: Currently we only support a single notification per recipient per (scope_and_id) combo, which means either 'organization', or 'project', or 'user'. If you were to set another one to the same recipient in the same scope, it overrides your previous notification to that recipient in that scope. If you then delete that notification it only deletes the single notification

File: docs/docs/cloud/cloud-cli.mdx
Comment (leoroese) [reply docs/docs/cloud/cloud-cli.mdx:98]: I was using general examples in order to showcase how the logic works, we only support a single job_run event right now but that can be changed. I did this because it didn't make a lot of sense to split up into job_run_succeeded, job_run_cancelled when we were setting the "status" as a filter option.

I almost wonder if we should split it into multiple events and not have "status" as a filter, but instead derive the "status" from the event. So instead of 
events: ["job_run"], status: ["succeeded", "failed"]

We had
events: ["job_run_succeeded", "job_run_failed"]

Initially I went with the first option to make it easier to specify the same statuses for different events so I could do something like `post this notification for all "job_run" and "project_create" that "succeeded" or "failed"` instead of
events: ["job_run_succeeded", "job_run_failed", "project_create_succeeded", "project_create_failed"]

This is also given the assumption that most of our events that trigger notifications will have a "status".

Comment (tayloramurphy) [reply docs/docs/cloud/cloud-cli.mdx:98]: Let's not change the model - I think it makes sense as is. All I'm looking for is an example of what folks would set (assuming most people like to copy paste things)

### Job Failures

To subscribe only to job failures you would set the filter to: 

```
example command
```

----

Though now that you mention it, this paragraph says we have the following events `job_created", "job_succeeded` but that doesn't match the filter listed here https://deploy-preview-7968--meltano.netlify.app/cloud/concepts/#supported-notification-filters

File: docs/docs/cloud/cloud-cli.mdx
Comment (tayloramurphy) [reply docs/docs/cloud/cloud-cli.mdx:127]: Let's add that as a comment in this section then. Good to document the limitations.

File: docs/docs/cloud/cloud-cli.mdx
Comment (tayloramurphy) [reply docs/docs/cloud/cloud-cli.mdx:92]: I like the idea of showing `Filter: N/A` or something like that. and then if there is a filter it's obvious what it is. 

Basically if someone has 10 notifications set up, how do they know which ones are for failed notifications or not?

Comment (tayloramurphy) [reply docs/docs/cloud/cloud-cli.mdx:99]: This looks good

File: docs/docs/cloud/cloud-cli.mdx
Comment (tayloramurphy) [inline docs/docs/cloud/cloud-cli.mdx:106]: @leoroese I'm running this exact example and am getting the following:

```bash
❯ meltano-cloud config notification set email --recipient taylor2@meltano.com --filter '{"events": ["job_run"], "status": ["failed"]}'
2023-08-22 16:51:54 [debug    ] Making Cloud CLI HTTP request  method=POST url=https://internal.api.meltano.cloud/notifications/v1/01GV0Z1VW5TSRF8BYX1J2V3WNT/cf6fc79ab63d9f69309059ecea8866416df7372ec1accbe1e103518730f24ca7/notification?entity_id=01H8D07597KR2Y474SSJG3CTT2
Error: . 

Please ensure you've correctly followed instructions at: https://docs.meltano.com/cloud/cloud-cli
If the issue persists, we'd be happy to help at: https://meltano.com/slack
```

File: docs/docs/cloud/cloud-cli.mdx
Comment (tayloramurphy) [reply docs/docs/cloud/cloud-cli.mdx:60]: Looks good

**Review by tayloramurphy — CHANGES_REQUESTED** [review]
I can't get the filters to work with the provided example.

File: docs/docs/cloud/cloud-cli.mdx
Comment (leoroese) [reply docs/docs/cloud/cloud-cli.mdx:106]: Taking a look

Comment (leoroese) [reply docs/docs/cloud/cloud-cli.mdx:106]: Screenshot 2023-08-22 at 5 19 21 PM

**Review by tayloramurphy — APPROVED** [review]
Approving based on async convos

File: src/meltano/cloud/cli/config.py
Comment (edgarrmondragon) [reply src/meltano/cloud/cli/config.py:606]: Yeah, not a blocker 👍

**Review by edgarrmondragon — APPROVED** [review]

**Pull request merged**
